### ISSUE TITLE
Question on IP Submission

### ISSUE THREAD

**Issue by agnesnatasya** [nus-cs2103-AY1920S2/forum#11]
Hi, I had submitted my IP by pushing the changes to my remote repo. 

I want to ask if I want to make a drastic change to my submission (this change involves change to my level 1), should I resubmit everything from level 1, or I just make a new commit and label it as Level 6. If I do it the second way, then my level 1-5 will have a different design with my level 5. 

Is the IP graded according to the design of the program or just its functionalities?

Also, for A-TextUITesting, I should replace all '..\src' and '..\bin' in the Duke tutorial file to the directory in my computer (such as Users\duke\src), am I correct? 
However, I can't find bin directory inside (Users\duke). Should I leave the  '..\bin'?

For A-Enums, may I know how to implement Enums in the IP?

**Comment by mario7lorenzo** [comment]
Hi there, I think I can only answer the last 2 questions since the first 2 questions are about the administration.

For the text UI testing, you don't need to replace the "../the_directory_name" since "../" means the parent of the current directory. So you don't need to modify anything.

For the enums, I assume you ask where should you implement Enums, instead of how to make it (?) If you really ask the latter, you can refer to [https://nus-cs2103-ay1920s2.github.io/website/schedule/week1/topics.html#key-exercise-show-priority-color](url).

If you ask for where you should implement enums, well you can think of the instances where the size of possibilities is small enough for you to implement enums, and you feel that you can minimize the risk of invalid things. For example, you can use Enum for your command keywords. It is easy to implement enums because of the possibilities of the keywords are small, and enums can prevent from invalid keywords as well.

**Comment by agnesnatasya** [comment]
Thank you for your answer!

But in the instruction of the tutorial for the UI testing, it says "Update the javac and java commands in the script to match the name/location of your main class."

And what should I add to the input.txt and Expected.txt, is it the messages entered to the bot duke like
"deadline return book /by Sunday"
"event project meeting /at Mon 2-4pm"
to the in.TXT? Also, should we commit and submit these input.txt, expected.txt, runtest.sh to the IP?

For my enums question, what I mean was the project asks us to implement Enums if applicable, however I am not sure why do we need Enums at all because I do not use Enums in my program, is it necessary or compulsory to use Enums in the project?

Thank you.

**Comment by chrisjwelly** [comment]
For input.txt and Expected.txt, I simply added a possible series of input a user can make.

I think for now, this can be anything we want? After all, one of the purposes of testing is to prevent regression; we want to prevent future changes to break functionalities we have now.

I did commit those files into my repo, but I am not very sure here either.

**Comment by zhuhanming** [comment]
Hi @agnesnatasya, for the tutorial on textUiTesting, if the entry point to your program is not via the Duke class, or if you changed the organisation of the folders, then you would need to update the default values provided in the `javac` and `java` lines. For example, my program starts from the `Main` class, instead of the default `Duke` class. I would thus change my `javac` to compile all `.java` files found in `src/main/java` and `java` to run `Main` instead of `Duke`. There is another issue opened in this repo that helps with the compilation of all `.java` files.

As for why you don't need to change the directories `../src` and `../bin`, as @mario7lorenzo mentioned, the reference `../` is **relative** to where this file is placed. It refers to the parent directory, in this case, the project root directory. This is desirable because you would want your project to be able to run no matter where it is cloned. Hardcoding the directory would only mean that it can only run on your machine, in the exact directory that it is stored at right now.

For your second question on `input.txt` and `EXPECTED.txt`, yes, your `input.txt` will contain all the commands given on separate lines (no need for any blank line between them). If you're using the built-in Java Scanner class in your programme, the `scanner.nextLine()` reads input line by line (based on the line breaks). The `EXPECTED.txt` would then contain what you would expect to be printed to your console (purely output). I believe we should commit all of these files to the repository.

For enums, I believe it is good practice to implement it for commands. As @mario7lorenzo mentioned, you know exactly what commands you are able to handle, and the number is quite small. Anyone trying to understand your code can easily look at it and immediately see what functionalities your program supports. Furthermore, as you can have methods with enums, using enums can actually help with abstraction. But I believe this is largely a personal choice, there may be other ways to implement enums further down the road.

**Comment by mario7lorenzo** [comment]
Ah ok. If you store your files in the src/main directory (you don’t change the default directory where the original Duke.java is located), you don’t need to change the shell script. Otherwise, you need to change it.

For the .txt files, you can put in your test cases in your input.txt, you can put the list of commands you want to test. For example, 
```
list
todo helloworld
done 2

```
The expected.txt file should be filled with the expected output, formatted properly. It must be formatted exactly the same as the output. Extra spaces and newlines can cause the testing to fail.

For the Enums, if you feel it is not necessary to use Enums, then you dont need to since it is stated “if applicable”

**Comment by agnesnatasya** [comment]
Thank you for the answers! 

I run to this cannot find symbol error when running my runtest.sh because they only compile my Duke.java and not the other classes. I noticed another post about compiling all files in a folder, but I think that is for the Windows version. Does anyone know how to do it for Mac?

For the expected.txt, it should be generated from our own output right? Then we will compare it with out.txt, which will be exactly the same with expected.txt? Or should we fill up expected.txt from the project website and copy their outputs? If I do the latter, the problem is that my indentation is different with the one in the website. They have 4 spaces, and 5 spaces for printing while I only have 3 spaces and 4 spaces. Is this a problem? Should I change all my outputs to have 4 and 5 spaces instead?

Thank you!

**Comment by mario7lorenzo** [comment]
Yeah. You need to compile all other classes as well. If your other classes are within the same directory as your Duke, you can change Duke.java to *.java in your shell script.

For your expected.txt, you should manually provide the output. For example, if you have “todo helloworld” in your input.txt, you should manually write your expected output e.g’ “Done! I have added helloworld to the list. You have 1 task currently”.

Yes, even single space and indentation will fail the test because it is basically use a “diff” command in your terminal which will try to find the exact difference, and indeed, it will detect different spaces and indentation.

Therefore, you should format the output that matches. So a suggestion from me is, just copy the template of the output first, then just modify the expected output for the subsequent tests corresponding to the input.

**Comment by zhuhanming** [comment]
From this StackOverflow post [here](https://stackoverflow.com/questions/6623161/javac-option-to-compile-all-java-files-under-a-given-directory-recursively)

You can use
```
if
  ! (
    find ../src/main/java -name "*.java" >sources.txt
    javac -cp ../src -Xlint:none -d ../bin @sources.txt
  )
then
```
for Linux/Mac. This solution will help to find all `.java` files, even inside subdirectories.

Alternatively, if all of your `.java` files are in the `../src/main/java` folder i.e. no subfolders, then you can also simply use this one-line solution:
```
javac -cp ../src -Xlint:none -d ../bin $(find ../src/main/java/* | grep .java);
```

I believe there is no need to replicate the output on the module website exactly - you should test your output against what you have programmed your Duke to output, and try to do so manually.

EDIT: Fixed the suggested answer above.

**Comment by okkhoy** [comment]
Good question(s) and better answers here! I like this thread! 😄 

I suspect the technical queries are answered here to a large extent thanks to @mario7lorenzo @zhuhanming and @chrisjwelly .

Let me answer the administrative questions

1. If you have changed the design which makes L1-L5 different from L6, **you don't need to** resubmit all other levels. Just tagging it as Level-6 is sufficient. You are getting a first-hand experience of a software project here; things change continuously, and you seem to be adapting well!

2. For Duke, we don't have the bandwidth to check your design; so we will not consider design aspects in the grading. Treat these exercises as:
  i. a refresher to your OOP concepts.
  ii. learn _some_ aspects of software engineering (like following basic coding standards, handling _some_ types of exceptions to make your code robust, building _some_ infrastructure for automated regression testing, etc.,)
  iii. a way to let personalize the product the way you want so that you can have fun doing this!

On a related note, grading is based on your attempt (checked by the automated scripts) and achieving the minimal required functionality (checked by the demo to tutors). So as long as the minimum bar is met, you can decide to stop wherever you want. If you are having fun and want to do more, I won't stop you (but doing more is not required!).


PS: note that _some_ is very subjective, look at the minimum bar mentioned for the levels to decide on the some! You will be doing all of this (hopefully) with much more rigor in your team projects!

**Comment by agnesnatasya** [comment]
Thank you for your answers everyone, it helped me a lot! Thank you for clarifying my administrative questions as well 😃
